File: windows/tray-app/RifeZPhoneBridge.Core/Protocol/FlatBufferHelloProtocol.cs

using Google.FlatBuffers;
using RifeZ.PhoneAudio.Control;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;


namespace RifeZPhoneBridge.Core.Protocol;

public static class FlatBufferHelloProtocol
{
    public static byte[] BuildHelloRequest(string clientName)
    {
        var builder = new FlatBufferBuilder(256);

        var clientNameOffset = builder.CreateString(clientName);

        HelloRequest.StartHelloRequest(builder);
        HelloRequest.AddClientName(builder, clientNameOffset);
        var helloOffset = HelloRequest.EndHelloRequest(builder);

        Envelope.StartEnvelope(builder);
        Envelope.AddMessageType(builder, MessageType.HelloRequest);
        Envelope.AddPayloadType(builder, Payload.HelloRequest);
        Envelope.AddPayload(builder, helloOffset.Value);
        var envelopeOffset = Envelope.EndEnvelope(builder);

        builder.Finish(envelopeOffset.Value, "RFZ1");
        return builder.SizedByteArray();
    }

    public static string ParseHelloResponse(byte[] payload)
    {
        var bb = new ByteBuffer(payload);
        var envelope = Envelope.GetRootAsEnvelope(bb);

        if (envelope.MessageType != MessageType.HelloResponse)
            throw new InvalidOperationException($"Unexpected message type: {envelope.MessageType}");

        var response = envelope.Payload<HelloResponse>();
        if (!response.HasValue)
            throw new InvalidOperationException("HelloResponse payload missing.");

        return response.Value.ReceiverName ?? string.Empty;
    }
}

File: android/receiver-app/app/src/main/java/com/rifez/phoneaudio/protocol/FlatBufferHelloProtocol.kt
package com.rifez.phoneaudio.protocol

import RifeZ.PhoneAudio.Control.Envelope
import RifeZ.PhoneAudio.Control.HelloRequest
import RifeZ.PhoneAudio.Control.HelloResponse
import RifeZ.PhoneAudio.Control.MessageType
import RifeZ.PhoneAudio.Control.Payload
import com.google.flatbuffers.FlatBufferBuilder
import java.nio.ByteBuffer

object FlatBufferHelloProtocol {

    fun parseHelloRequest(payload: ByteArray): String {
        val bb = ByteBuffer.wrap(payload)
        val envelope = Envelope.getRootAsEnvelope(bb)

        require(envelope.messageType() == MessageType.HelloRequest) {
            "Unexpected message type: ${envelope.messageType()}"
        }

        val helloRequest = HelloRequest()
        val payloadOk = envelope.payload(helloRequest)
        require(payloadOk != null) { "HelloRequest payload missing" }

        return helloRequest.clientName() ?: ""
    }

    fun buildHelloResponse(receiverName: String): ByteArray {
        val builder = FlatBufferBuilder(256)

        val receiverNameOffset = builder.createString(receiverName)

        HelloResponse.startHelloResponse(builder)
        HelloResponse.addReceiverName(builder, receiverNameOffset)
        val helloResponseOffset = HelloResponse.endHelloResponse(builder)

        Envelope.startEnvelope(builder)
        Envelope.addMessageType(builder, MessageType.HelloResponse)
        Envelope.addPayloadType(builder, Payload.HelloResponse)
        Envelope.addPayload(builder, helloResponseOffset.toLong().toInt())
        val envelopeOffset = Envelope.endEnvelope(builder)

        builder.finish(envelopeOffset, "RFZ1")
        return builder.sizedByteArray()
    }
}